Handle order of PhysicalOffsetFrame

aseth1 opened the pull request:
Addresses #1255.

Only two significant changes:
 1. Introduced a test case in `testFrames` to produce a failure when `PhysicalOffsetFrame`s are added to the `Model` in an order that does not correspond to the Multibody topology.
 2. `Model::extendConnectToModel` now traverses its `PhysicalOffsetFrame` sub-components so that its parent `PhysicalFrame`s are queued up first to establish proximal to distal order. 

Should be a short review.

chrisdembia commented on OpenSim/Simulation/Model/Model.cpp:
```diff
@@ -769,11 +769,31 @@ void Model::extendConnectToModel(Model &model)
+        // PhysicalOffsetFrame can be listed in any order and may be attached
+        // to any other PhysicalOffsetFrame, so we need to find their parent(s)
+        // in the tree and add them first.
+        pof.connect(*this);
```
This will cause all subcomponents of `pof` to be connected, right? Do you want that, or do you specifically want to connect the `parent` connector?

chrisdembia replied:
Okay; if performance becomes an issue, we can revisit later.

chrisdembia commented on OpenSim/Simulation/Model/Model.cpp:
```diff
@@ -769,11 +769,31 @@ void Model::extendConnectToModel(Model &model)
+        pof.connect(*this);
+        const PhysicalOffsetFrame* parentPof =
+            dynamic_cast<const PhysicalOffsetFrame*>(&pof.getParentFrame());
+        std::vector<const PhysicalOffsetFrame*> parentPofs;
```
This is fine as-is. But: you are using this as a stack; I wonder if there is any benefit to using `std::stack`?

aseth1 commented on OpenSim/Simulation/Test/testFrames.cpp:
```diff
@@ -332,6 +341,45 @@ void testPhysicalOffsetFrameOnBodySerialize()
+
+    SimTK::Transform X_RO;
+    X_RO.setP(SimTK::Vec3(0.1, 0.2, 0.3));
+    X_RO.updR().setRotationFromAngleAboutAxis(SimTK::Pi / 4.0, SimTK::ZAxis);
```
No significance whatsoever. I *borrrowed* the code block from the `testPhysicalOffsetFrameOnBodySerialize()` case earlier in `testFrames`. The outcome of the transform is not being tested, so it doesn't matter what value it has..

aseth1 commented on OpenSim/Simulation/Test/testFrames.cpp:
```diff
@@ -332,6 +341,45 @@ void testPhysicalOffsetFrameOnBodySerialize()
+    PhysicalOffsetFrame* offsetFrameProximal = new PhysicalOffsetFrame();
+    offsetFrameProximal->setName("offsetFrameProximal");
+    offsetFrameProximal->setOffsetTransform(~X_RO);
+    pendulum.addComponent(offsetFrameProximal);
```
I will add it.

aymanhab commented:
@aseth1 My main issue is the knowledge about the function of various Components/Order that's baked into Model::extendConnectToModel which may make it harder to interpret/extend in the future (e.g. if we introduce a new type of Frame will this code need to change? how? What about new types of Components?) If you believe there's no way around it then we can live with it.

One other high level question before diving into the details. Have you considered using an actual Simbody 'Body' underneath PhysicalOffsetFrame? This will delegate the ordering to one commonplace/ tree-builder on the Simbody side and may make all the issues with SimbodyEngine go away but I'm not sure what's the price/overhead.

aseth1 commented:
>  My main issue is the knowledge about the function of various Components/Order that's baked into Model::extendConnectToModel which may make it harder to interpret/extend in the future (e.g. if we introduce a new type of Frame will this code need to change? how? What about new types of Components?)

It is a major concern for sure and ideally a Model or any top-level component would not have special knowledge and handling of any of its constituents.  However, the main task of an OpenSim Model is to translate between the world of the modeler/user (biomechanist, movement scientist, etc..) to something that can be solved (the MultibodySystem). A MultibodySystem requires a well-defined tree topology whereas a Model (ideally) does not. We have made a lot of progress in that direction and we can handle any order of bodies and joints, for example. Nonetheless, it is the Model's responsibility to construct the appropriate topology for the user's model, so a Model has to know what Components define the topology of a MultibodySystem and these primarily Joints and the rigid bodies they connect. A `PhysicalOffsetFrame` then is an extension of a rigid body and provides convenience for locating attachments and measuring motion in easier to interpret reference frames. As an extension of a rigid body, a `PhysicalOffsetFrame` is part of the tree and has to be treated as such.  

I think we have the rigid bodies (aka `PhysicalFrame`s) covered. They can and should be embedded wherever a rigid connection location is needed. As for new types of Frames, like representing the center of pressure or the center of mass are not part of the Multibody topology and are computed based on the system topology, so these would not require special handling. 

In the case of new Components, new Components should avoid (and perhaps we should prevent it completely) creating MobiliizedBodies themselves. Joints are the OpenSim abstraction for adding dgerees-of-freedom and new Components that need additional dofs and need new attachments should use Joints and PhysicalFrames internally. The exception to this is creating new types of Joint components, in which case deriving from Joint ensures that it is appropriated integrated into the System topology.

> One other high level question before diving into the details. Have you considered using an actual Simbody 'Body' underneath PhysicalOffsetFrame? This will delegate the ordering to one commonplace/ tree-builder on the Simbody side and may make all the issues with SimbodyEngine go away but I'm not sure what's the price/overhead.

It it is a good point and one we discussed a while ago when considering `PhysicalOffsetFrame` as a welded massless Body. The main issues with doing that, as I recall,  were: 1) complete redundancy with `WeldJoint`. 2) `MultibodySystem` size and book-keeping increases with the number of SimTK::Bodies (even though the number of ODEs does not change) whereas adding a `PhysicalOffsetFrame` has no impact on the System computations until it is asked to calculate its transform. In the typical scenario of connecting Bodies with Joints it would introduce Welds for every parent and child frame, where as currently they have no impact since they get baked into the creation of the underlying `MobilizedBody`; 3) asking the `MultibodySystem` for the number of bodies would not even remotely correspond to the number of Bodies in the Model, and could potentially double, triple or even quadruple the number of underlying bodies.

aseth1 commented:
@aymanhab sorry for the lengthy response, but I hope it addressed your concerns.

@chrisdembia I added comments to describe what is being tested and how.

aymanhab commented:
Thanks for the explanation @aseth1  I had one small question above, other than that LGTM.

chrisdembia commented on OpenSim/Simulation/Model/Model.cpp:
```diff
@@ -769,11 +769,31 @@ void Model::extendConnectToModel(Model &model)
+            parentPof =
+                dynamic_cast<const PhysicalOffsetFrame*>(&parentPof->getParentFrame());
+        }
+        while (parentPofs.size()) {
```
I think you might be referring to the while-loop above, the one that involves the dynamic cast. The one below should certainly terminate, as it pops off each element.

aseth1 replied:
The rationale is that moving inward the final PhysicalOffsetFrame has to be connected to a Body or Ground, which would cause the dynamic_cast to return nullptr and terminate the loop. That should work for any well formed tree, but I see now that it is certainly possible to *connect* PhysicalOffsetFrame's such that they form loops and that should be prevented. I am not sure that should happen here, but I agree a check is needed.

aseth1 commented:
@aymanhab thanks for pointing out the potential for loops of `PhysicalOffsetFrames` which is indeed an error case that wasn't being checked. I've added the check and included an Exception when a loop is detected. I also added two more conditions in `testFrames.cpp` to verify that the exception is thrown when expected. 
@chrisdembia feel free to take a look too. Thanks!

aseth1 commented:
@aymanhab @chrisdembia CI finally completed successfully.

aymanhab commented:
Thanks @aseth1 LGTM. Feel free to merge when tests pass.

aseth1 commented:
:champagne: cheers @chrisdembia + @aymanhab
